lwt/happy_eyeballs_lwt.ml

```diff
@@ -152,21 +166,34 @@ let rec timer t =
   loop ()
 
 let ctr = ref 0
+let error_msgf fmt = Fmt.kstr (fun msg -> Error (`Msg msg)) fmt
 
-let create ?(happy_eyeballs = Happy_eyeballs.create (now ())) ?dns ?(timer_interval = Duration.of_ms 10) () =
-  let dns =
-    Option.value ~default:
-      (let timeout = Happy_eyeballs.resolve_timeout happy_eyeballs in
-       Dns_client_lwt.create ~timeout ())
-      dns
-  in
+let getaddrinfo record domain_name =
+  let open Lwt.Infix in
+  let getaddrinfo_option = match record with
+    | `A -> [ Unix.AI_FAMILY Unix.PF_INET ]
+    | `AAAA -> [ Unix.AI_FAMILY Unix.PF_INET6 ] in
+  let getaddrinfo_option = Unix.AI_SOCKTYPE Unix.SOCK_STREAM :: getaddrinfo_option in
+  Lwt.catch
+    (fun () -> Lwt_unix.getaddrinfo (Domain_name.to_string domain_name) "" getaddrinfo_option)
+    (fun exn -> Lwt.return []) >|= function
+  | [] -> error_msgf "%a not found" Domain_name.pp domain_name
+  | addrs ->
+    let set = List.fold_left (fun set { Unix.ai_addr; _ } -> match ai_addr with
+      | Unix.ADDR_INET (inet_addr, _) -> Ipaddr.Set.add (Ipaddr_unix.of_inet_addr inet_addr) set
+      | Unix.ADDR_UNIX _ -> set)
+      Ipaddr.Set.empty addrs in
+    Ok set
+
+let create ?(happy_eyeballs = Happy_eyeballs.create (now ())) ?(getaddrinfo= getaddrinfo)
+  ?(timer_interval = Duration.of_ms 10) () =
```

Comment:
Since we add a `?getaddrinfo` argument I wonder if we really need `inject`? Is it desirable to switch implementation at a later point?

Reply:
For the MirageOS purpose yes, we must create a dummy implementation of `getaddrinfo` and inject it later when we create the DNS instance (again, it's about depedency cycle).